### ISSUE TITLE
Identify when partitioning is completed

### ISSUE THREAD

**Issue by awoods187** [cockroachdb/cockroach#31551]
It is currently unclear when a partitioning event has completed in CockroachDB. 

**Describe the solution you'd like**
I would like to be able to understand when partitioning has been completed so that we can test performance at the end of partitioning. 

**Describe alternatives you've considered**
According to @danhhz scatter used to not complete until partitioning completed but this may not be so anymore. 

**Additional context**
The cluster should still be able to serve traffic while partitioning is occurring, and, ideally, partitioning would not impact performance.

**Comment by danhhz** [comment]
We talked about this a bit when building partitioning. Here's what I remember:

The biggest question here is the UX of how the user does this waiting. Either a partition command or a zone config command can cause work to be done. For example, if an initial partitioning is set, nothing happens until a zone config targets those partitions. If the values in those partitions is later changed, then the repartitioning causes the work to be done. So, the simplest thing is to make this system wait for both partition and zone config commands (perhaps this becomes a sql option, with either a default of wait or nowait).

A slightly more flexible alternative is to add a command to wait on a set of databases, tables, or partitions to be balanced.

At one point, scatter worked by forcing the targeted replicas through the rebalance queue until it said there was no more work to be done. This made it happen to work for the second idea I described. But scatter has been tweaked a bunch since then and I haven't kept track of how it works now.

**Comment by a-robinson** [comment]
> According to @danhhz scatter used to not complete until partitioning completed but this may not be so anymore.

Has anyone verified that one way or the other? I suspect it doesn't wait fully given that it only forces targeted replicas through the rebalance queue until the lease gets transferred, which can happen if the leaseholder decides that it should be removed from the range (even if there's more rebalancing that should be done after that).

A "wait for rebalancing to finish" command could be implemented on top of Scatter very easily by having `AdminScatter` return whether it definitely finished all its desired rebalancing, and if it didn't then the caller would have to call it against the new leaseholder.

**Comment by knz** [comment]
I also think that Andy's concern is how to *observe* the thing has completed, and not to *wait* on it.

Perhaps a job entry with a suitable "status" field would be sufficient. This would be a significantly lighter technical problem to solve.

@awoods187 correct me if I'm wrong.

**Comment by danhhz** [comment]
I've been tying them together because I personally don't see how observing it is easier than waiting on it, but happy to be proven wrong.

**Comment by knz** [comment]
So my concern is that of UX: there **must** be a job with a status updated properly. This because if the client that issued the SCATTER loses their connection inadvertently, I still want them to be able to know when the operation completes. A job will help them with that.

**Comment by danhhz** [comment]
Makes sense. If we do that, we'd need jobs for both partitioning/repartitioning as well as for zone config updates that alter a constraint (because both can caused partitioned data to move). From a UX perspective, I don't see any reason why the latter should be specific to partitioning, it should work for any zone config updates that move data (which has the nice side-benefit of being easier to implement).

**Comment by petermattis** [comment]
Discussing with @nvanbenschoten today in the context of automated benchmarking. Nathan feels he can visually inspect a cluster and determine if it has "stabilized" following splitting/scattering/partitioning. Our thinking is that we can automate this visualize inspection by grabbing a few different timeseries metrics. Nathan will probably be experimenting with this in the near term and any success (or failure) can inform this issue.

**Comment by knz** [comment]
This is still relevant as per 21.1.

cc @ajstorm  - do we have another tracking issue for this already?

Also this connects to a convo we had yesterday with @aayushshah15 and @irfansharif about knowing when a zone config is fully applied.

**Comment by ajstorm** [comment]
@knz we have an epic around observability for multi-region operations (which would cover the partitioning changes as a result of moving to/from REGIONAL BY ROW), but I'm not sure if that will also cover the general partitioning case.  

@awoods187  thoughts?

**Comment by awoods187** [comment]
I think in an ideal world you'd be able to observe the primitives of partitioning so that you can know when they completed. I think this is a much lower priority now that we will be using the new multi region abstractions. I agree with Adam that we will be planning to step up our observability of those abstractions and we can likely close this in favor of those efforts.

**Comment by knz** [comment]
still relevant
